Sources/Controllers/2.0/OAMapSettingsWikipediaScreen.mm

```diff
@@ -0,0 +1,625 @@
+#import "OAWikipediaPlugin.h"
+#import "Localization.h"
+#import "OAColors.h"
+#import "OATableViewCustomFooterView.h"
+#import "OADividerCell.h"
+#import "OASettingSwitchCell.h"
+#import "OAIconTitleValueCell.h"
+#import "OAPOIFiltersHelper.h"
+#import "OAMapSettingsWikipediaLanguagesScreen.h"
+#import "OAWikiArticleHelper.h"
+#import "OAIAPHelper.h"
+#import "OARootViewController.h"
+#import "OAAutoObserverProxy.h"
+#import "OAPluginPopupViewController.h"
+#import "OAManageResourcesViewController.h"
+
+#define kCellTypeMap @"MapCell"
+
+static const NSInteger visibilitySection = 0;
+static const NSInteger languagesSection = 1;
+static const NSInteger availableMapsSection = 2;
+
+typedef OsmAnd::ResourcesManager::ResourceType OsmAndResourceType;
+
+@interface OAMapSettingsWikipediaScreen () <OAWikipediaScreenDelegate>
+
+@end
+
+@implementation OAMapSettingsWikipediaScreen
+{
+    OsmAndAppInstance _app;
+    OAIAPHelper *_iapHelper;
+    OAMapViewController *_mapViewController;
+
+    OAWikipediaPlugin *_wikiPlugin;
+    NSArray<OARepositoryResourceItem *> *_mapItems;
+
+    OAAutoObserverProxy* _downloadTaskProgressObserver;
+    OAAutoObserverProxy* _downloadTaskCompletedObserver;
+    OAAutoObserverProxy* _localResourcesChangedObserver;
+
+    NSObject *_dataLock;
+    BOOL _wikipediaEnabled;
+    NSArray<NSArray <NSDictionary *> *> *_data;
+}
+
+@synthesize settingsScreen, tableData, vwController, tblView, title, isOnlineMapSource;
+
+- (id)initWithTable:(UITableView *)tableView viewController:(OAMapSettingsViewController *)viewController
+{
+    self = [super init];
+    if (self)
+    {
+        _app = [OsmAndApp instance];
+        _iapHelper = [OAIAPHelper sharedInstance];
+        settingsScreen = EMapSettingsScreenWikipedia;
+        vwController = viewController;
+        tblView = tableView;
+        _wikiPlugin = (OAWikipediaPlugin *) [OAPlugin getPlugin:OAWikipediaPlugin.class];
+        _dataLock = [[NSObject alloc] init];
+        _wikipediaEnabled = [[OAPOIFiltersHelper sharedInstance] isTopWikiFilterSelected];
+        _mapViewController = [OARootViewController instance].mapPanel.mapViewController;
+        [self setupView];
+        [self generateData];
+    }
+    return self;
+}
+
+- (void)dealloc
+{
+    if (_downloadTaskProgressObserver)
+    {
+        [_downloadTaskProgressObserver detach];
+        _downloadTaskProgressObserver = nil;
+    }
+    if (_downloadTaskCompletedObserver)
+    {
+        [_downloadTaskCompletedObserver detach];
+        _downloadTaskCompletedObserver = nil;
+    }
+    if (_localResourcesChangedObserver)
+    {
+        [_localResourcesChangedObserver detach];
+        _localResourcesChangedObserver = nil;
+    }
+}
+
+- (void)initData
+{
+}
+
+- (void)setupView
+{
+    _downloadTaskProgressObserver = [[OAAutoObserverProxy alloc] initWith:self withHandler:@selector(onDownloadTaskProgressChanged:withKey:andValue:) andObserve:_app.downloadsManager.progressCompletedObservable];
```

Comment:
Observers have nothing to do with the view setup and should be initialized separately in the init method.